File: src/suffixarray_search.py
import time
import gzip
import sys
import iv2py as iv

def load_fasta(filepath):
    sequences = []
    current_seq = []
    
    open_func = gzip.open if filepath.endswith('.gz') else open
    
    with open_func(filepath, 'rt') as f:
        for line in f:
            line = line.strip()
            if line.startswith('>'):
                if current_seq:
                    sequences.append(''.join(current_seq))
                    current_seq = []
            else:
                current_seq.append(line)
        
        if current_seq:
            sequences.append(''.join(current_seq))
    
    return sequences

class SuffixArraySearch:    
    def __init__(self, reference):
        self.reference = reference
        self.construction_time = 0
        
        start = time.time()
        self.sa = iv.create_suffixarray(reference)
        self.construction_time = time.time() - start
    
    def search(self, query):
        results = []
        query_len = len(query)
        n = len(self.sa)
        
        # Binary search for start of matching range
        left, right = 0, n - 1
        start_idx = n
        
        while left <= right:
            mid = (left + right) // 2
            pos = self.sa[mid]
            
            if pos + query_len <= len(self.reference):
                cmp_result = self.reference[pos:pos + query_len]
                
                if cmp_result >= query:
                    start_idx = mid
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                right = mid - 1
        
        # Binary search for end of matching range
        left, right = 0, n - 1
        end_idx = -1
        
        while left <= right:
            mid = (left + right) // 2
            pos = self.sa[mid]
            
            if pos + query_len <= len(self.reference):
                cmp_result = self.reference[pos:pos + query_len]
                
                if cmp_result <= query:
                    end_idx = mid
                    left = mid + 1
                else:
                    right = mid - 1
            else:
                right = mid - 1
        
        # Collect all matching positions
        if start_idx <= end_idx:
            for i in range(start_idx, end_idx + 1):
                pos = self.sa[i]
                if pos + query_len <= len(self.reference):
                    if self.reference[pos:pos + query_len] == query:
                        results.append(pos)
        
        return results

def benchmark_query_counts(query_count):
    reference_file = "hg38_partial.fasta.gz"
    references = load_fasta(reference_file)
    reference = references[0]
    print(f"Reference length: {len(reference):,} bp")
    
    queries_100 = load_fasta("illumina_reads_100.fasta.gz")
    
    test_queries = queries_100[:query_count]
    print(f"{len(test_queries):,} queries of length 100 bp\n")    
    # Build suffix array ONCE
    searcher = SuffixArraySearch(reference)
    print(f"Construction time for suffix array: {searcher.construction_time:.4f}s\n")
    
    # Search all queries
    start = time.time()
    total_hits = 0
    for query in test_queries:
        hits = searcher.search(query)
        total_hits += len(hits)
    search_time = time.time() - start
    
    # Results
    print(f"Results for {query_count:,} queries: ")
    print(f"Construction time: {searcher.construction_time:.4f}s")
    print(f"Search time:       {search_time:.4f}s")
    print(f"Total time:        {searcher.construction_time + search_time:.4f}s")
    print(f"Hits found:        {total_hits:,}")

def benchmark_query_lengths():
    reference_file = "hg38_partial.fasta.gz"
    references = load_fasta(reference_file)
    reference = references[0]
    print(f"Reference length: {len(reference):,} bp\n")
    
    searcher = SuffixArraySearch(reference)
    print(f"Construction time of suffix array: {searcher.construction_time:.4f}s\n")
    
    query_files = {
        40: "illumina_reads_40.fasta.gz",
        60: "illumina_reads_60.fasta.gz",
        80: "illumina_reads_80.fasta.gz",
        100: "illumina_reads_100.fasta.gz"
    }
    
    num_queries = 10000
    results = []

    for length in sorted(query_files.keys()):
        filename = query_files[length]
        print(f"\nQuery length: {length} bp")
        
        queries = load_fasta(filename)
        test_queries = queries[:min(num_queries, len(queries))]
        
        start = time.time()
        total_hits = 0
        for query in test_queries:
            hits = searcher.search(query)
            total_hits += len(hits)
        search_time = time.time() - start
        
        results.append({
            'length': length,
            'queries': len(test_queries),
            'search_time': search_time,
            'hits': total_hits
        })
        
        print(f"  Queries:     {len(test_queries):,}")
        print(f"  Search time: {search_time:.4f}s")
        print(f"  Hits found:  {total_hits:,}")
        print(f"  Time/query:  {search_time/len(test_queries)*1000:.4f}ms")

if __name__ == "__main__":
    if len(sys.argv) > 1:
        # Qn 4) Query count mode (for time and memory benchmarking with /usr/bin/time -v)
        query_count = int(sys.argv[1])
        benchmark_query_counts(query_count)
    else:
        # Qn 5) Query length mode (time benchmarking only)
        benchmark_query_lengths()

File: src/suffixarray_search.cpp
#include <divsufsort.h>
#include <sstream>
#include <algorithm>
#include <chrono>
#include <iomanip>
#include <vector>

#include <seqan3/alphabet/nucleotide/dna5.hpp>
#include <seqan3/argument_parser/all.hpp>
#include <seqan3/core/debug_stream.hpp>
#include <seqan3/io/sequence_file/all.hpp>
#include <seqan3/search/fm_index/fm_index.hpp>
#include <seqan3/search/search.hpp>

// Compare suffix starting at position 'suffix_pos' with query
int compareSuffix(std::vector<seqan3::dna5> const& reference,
                  saidx_t suffix_pos,
                  std::vector<seqan3::dna5> const& query) {
    size_t cmp_len = std::min(query.size(), reference.size() - suffix_pos);
    
    for (size_t i = 0; i < cmp_len; ++i) {
        if (reference[suffix_pos + i] < query[i]) {
            return -1;
        } else if (reference[suffix_pos + i] > query[i]) {
            return 1;
        }
    }
    
    if (cmp_len == query.size()) {
        return 0;
    }
    
    return -1;
}

// Binary search to find the leftmost occurrence
saidx_t binary_search_left(std::vector<saidx_t> const& sa, 
                            std::vector<seqan3::dna5> const& reference,
                            std::vector<seqan3::dna5> const& query) {
    saidx_t left = 0;
    saidx_t right = sa.size();
    
    while (left < right) {
        saidx_t mid = left + (right - left) / 2;
        int cmp = compareSuffix(reference, sa[mid], query);
        
        if (cmp < 0) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    
    return left;
}

// Binary search to find the rightmost occurrence
saidx_t binary_search_right(std::vector<saidx_t> const& sa,
                             std::vector<seqan3::dna5> const& reference,
                             std::vector<seqan3::dna5> const& query) {
    saidx_t left = 0;
    saidx_t right = sa.size();
    
    while (left < right) {
        saidx_t mid = left + (right - left) / 2;
        int cmp = compareSuffix(reference, sa[mid], query);
        
        if (cmp <= 0) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    
    return left;
}

struct BenchmarkResult {
    size_t num_queries;
    size_t query_length;
    double construction_time;
    double search_time;
    double total_time;
    size_t hits;
};

int main(int argc, char const* const* argv) {
    seqan3::argument_parser parser{"suffixarray_search", argc, argv, seqan3::update_notifications::off};

    parser.info.author = "SeqAn-Team";
    parser.info.version = "1.0.0";

    auto reference_file = std::filesystem::path{};
    parser.add_option(reference_file, '\0', "reference", "path to the reference file");

    auto query_file = std::filesystem::path{};
    parser.add_option(query_file, '\0', "query", "path to the query file");

    auto number_of_queries = size_t{100};
    parser.add_option(number_of_queries, '\0', "query_ct", "number of query, if not enough queries, these will be duplicated");

    try {
         parser.parse();
    } catch (seqan3::argument_parser_error const& ext) {
        seqan3::debug_stream << "Parsing error. " << ext.what() << "\n";
        return EXIT_FAILURE;
    }

    auto total_start = std::chrono::high_resolution_clock::now();

    // loading our files
    auto reference_stream = seqan3::sequence_file_input{reference_file};
    auto query_stream     = seqan3::sequence_file_input{query_file};

    // read reference into memory
    std::vector<seqan3::dna5> reference;
    for (auto& record : reference_stream) {
        auto r = record.sequence();
        reference.insert(reference.end(), r.begin(), r.end());
    }

    // read query into memory
    std::vector<std::vector<seqan3::dna5>> queries;
    for (auto& record : query_stream) {
        queries.push_back(record.sequence());
    }

    size_t query_length = queries.empty() ? 0 : queries[0].size();

    // Check if we have enough queries
    if (queries.size() < number_of_queries) {
        seqan3::debug_stream << "Error: Not enough queries in file. ";
        seqan3::debug_stream << "Requested: " << number_of_queries << ", Available: " << queries.size() << "\n";
        return EXIT_FAILURE;
    }

    std::cout << "Processing " << number_of_queries << " queries (length " << query_length << " bp)...\n";
    // Array that should hold the future suffix array
    std::vector<saidx_t> suffixarray;
    suffixarray.resize(reference.size());

    // Start timing - SA construction
    auto start_sa = std::chrono::high_resolution_clock::now();

    // Construct suffix array using libdivsufsort
    sauchar_t const* str = reinterpret_cast<sauchar_t const*>(reference.data());
    divsufsort(str, suffixarray.data(), reference.size());

    auto end_sa = std::chrono::high_resolution_clock::now();
    double sa_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_sa - start_sa).count() / 1000.0;

    std::cout << "  Construction time: " << std::fixed << std::setprecision(4) << sa_time << "s\n";

    // Start timing - search
    auto start_search = std::chrono::high_resolution_clock::now();

    size_t total_matches = 0;
    for (auto& q : queries) {
        saidx_t left = binary_search_left(suffixarray, reference, q);
        saidx_t right = binary_search_right(suffixarray, reference, q);
        total_matches += (right - left);
    }

    auto end_search = std::chrono::high_resolution_clock::now();
    double search_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_search - start_search).count() / 1000.0;

    auto total_end = std::chrono::high_resolution_clock::now();
    double total_time = std::chrono::duration_cast<std::chrono::milliseconds>(total_end - total_start).count() / 1000.0;

    std::cout << "  Search time: " << std::fixed << std::setprecision(4) << search_time << "s\n";
    std::cout << "  Total time: " << std::fixed << std::setprecision(4) << total_time << "s\n";
    std::cout << "  Hits found: " << total_matches << "\n\n";

    return 0;
}